feature: Add HKCDB and HICDB for fetching standing orders

code-chris opened the pull request:
- Added HISPAS Segment for fetching supported sepa-pain message formats
- Added HICDBS Segment for detecting supported HICDB version
- Added Parsing for XML responses
- Added HKCDB and HICDB Segments for fetching standing orders for a account.

Prior99 commented on packages/fints/package.json:
```diff
@@ -75,6 +75,7 @@
     "dependencies": {
         "bind-decorator": "^1.0.11",
         "date-fns": "^1.29.0",
+        "fast-xml-parser": "^3.12.12",
```
You added this library but I can't see the yarn.lock file to have changed. Can you please commit it as well?

Prior99 commented on packages/fints/src/parse.ts:
```diff
@@ -37,4 +39,24 @@ export const Parse = {
+     *
+     * @return The parsed object.
+     */
+    xml<T>(str: string): T {
```
This should have the return type `unknown` and a function checking for the correct type should be added instead:

```
function isWhatever(arg: unknown): arg is Whatever {
    // Perform validation of type here.
}
```

Prior99 commented on packages/fints/src/segments/hicdb.ts:
```diff
@@ -0,0 +1,60 @@
+        ] = input;
+
+        const jsonMessage: CustomerCreditTransferInitiationV03 =
+            Parse.xml<document>(sepaMessage).Document.CstmrCdtTrfInitn;
```
Is it always guaranteed that `jsonMessage` has this type? Does the parser perform schema validation?

code-chris replied:
I assumed that. No, the library does no schema validation.

Prior99 commented on packages/fints/src/segments/hicdb.ts:
```diff
@@ -0,0 +1,60 @@
+            orderDay: Parse.num(orderDay),
+            lastOrderDate: lastOrder ? Parse.date(lastOrder) : null,
+            creationDate: Parse.date(jsonMessage.GrpHdr.CreDtTm),
+            debitorName: instructionInfo.Dbtr.Nm,
```
If we would use an object here, the response would stay compatible across more functions:

```
debitor: {
    name: instructionInfo.Dbtr.Nm,
    iban: instructionInfo.DbtrAcct.Id.IBAN,
    bic: instructionInfo.DbtrAgt.FinInstnId.BIC,
},
creditor: {
    name: creditTransaction.Cdtr.Nm,
    iban: creditTransaction.CdtrAcct.Id.IBAN,
    bic: creditTransaction.CdtrAgt.FinInstnId.BIC,
}
```

Prior99 commented on packages/fints/src/segments/hkcdb.ts:
```diff
@@ -0,0 +1,35 @@
+        const { iban, bic } = account;
+        const pain0101: string = painFormats.find(x => x.startsWith("urn") && x.indexOf("pain.001.001.03") !== -1);
+        const pain0103: string = painFormats.find(x => x.startsWith("urn") && x.indexOf("pain.001.003.03") !== -1);
+        return [
```
Please throw an error here if no pain message was found.

code-chris replied:
Yes, there should be an error in that case. Any tip how I can mock this "painFormats" array in the `test-pin-tan-client.ts`? I tried with this line, but it doesn't work...
```
jest.spyOn(Response.prototype, "painFormats", "get").mockReturnValue(["pain.001.003.03"]);
```

Prior99 commented on packages/fints/src/types.ts:
```diff
@@ -152,6 +152,68 @@ export interface SEPAAccount {
+    /**
+     * The name of the debitor.
+     */
+    debitorName: string;
```
See my comment in file `packages/fints/src/segments/hicdb.ts`

Prior99 commented on packages/fints/typings/pain.001.001.03.d.ts:
```diff
@@ -0,0 +1,390 @@
+import * as Primitive from './xml-primitives';
```
I guess this wile and the other `.d.ts` file were autogenerated. Instead of checking in autogenerated code, please add a yarn command for generating them, perhaps into a directory `generated` in the package's root directory and add it to the Makefile or as a `prebuild` step. This way, we don't have to maintain generated code, can reproduce this and stay up-to-date. Please feel freee to ask if you need help with this :)

code-chris replied:
Hm, I agree with you. They were generated with [cxsd](https://www.npmjs.com/package/cxsd). The problem is, I might change a small number of definitions (currently working on that). E.g. the `_exists` or `_namespace` variables in `BaseType` are internal and should be optional. Another thing is, that there are fields of type `Date` or `number`, which is logically correct, but in the parsed JSON it is still a `string`. So it could be difficult to automate this completely... However, a manually run-command to update could be useful.

Prior99 replied:
Perhaps we could run `patch` on it after generating it if there is no other way?

code-chris commented:
I removed the generated files and replaced them with a prebuild step for now. I didn't found a really good solution to preprocess them right now, except multiple regex replacements.

Prior99 commented:
Hey, I have not forgotten about this, I am just busy as hell. I will try to go through this ob the weekend. I am sorry for the inconvenience.
